rolling-shutter/collator/collator.go

```diff
@@ -101,25 +106,55 @@ func Run(ctx context.Context, cfg config.Config) error {
 		return err
 	}
 
-	c := collator{
-		Config: cfg,
-
-		l1Client:  l1Client,
-		l2Client:  l2RPCClient,
-		contracts: contracts,
-		p2p: p2p.New(p2p.Config{
-			ListenAddrs:    cfg.ListenAddresses,
-			BootstrapPeers: cfg.CustomBootstrapAddresses,
-			PrivKey:        cfg.P2PKey,
-		}),
-		batcher:   btchr,
-		dbpool:    dbpool,
-		submitter: submitter,
-	}
-	c.submitter.collator = &c
+	c.l1Client = l1Client
+	c.l2Client = l2RPCClient
+	c.contracts = contracts
+	c.p2p = p2p.New(p2p.Config{
+		ListenAddrs:    cfg.ListenAddresses,
+		BootstrapPeers: cfg.CustomBootstrapAddresses,
+		PrivKey:        cfg.P2PKey,
+	})
+	c.batcher = btchr
+	c.dbpool = dbpool
+	c.submitter = submitter
+	c.submitter.collator = c
 	c.setupP2PHandler()
 
-	return c.run(ctx)
+	httpServer := &http.Server{
+		Addr:              c.Config.HTTPListenAddress,
+		Handler:           c.setupRouter(),
+		ReadHeaderTimeout: 10 * time.Second,
+	}
+
+	newSignal := func(signalName string, handler shdb.SignalHandler) shdb.SignalFunc {
+		sig, loop := shdb.NewSignal(ctx, signalName, handler)
+		runner.Go(loop)
+		return sig
+	}
+	c.signals.newDecryptionTrigger = newSignal(newDecryptionTrigger, c.sendDecryptionTriggers)
+	c.signals.newDecryptionKey = newSignal(newDecryptionKey, c.submitter.submitBatch)
+	c.signals.newBatchTx = newSignal(newBatchtx, c.submitter.submitBatchTxToSequencer)
+
+	runner.Go(httpServer.ListenAndServe)
+	runner.Go(func() error {
+		<-ctx.Done()
```

Comment:
ok, this can't be a `runner.Defer` because if there are successive defer functions this would block and not execute in parallel.

I'm a little bit on the fence about this `runnner.Go` usage...
Most of this code looks very similar to before, but makes it more complicated to understand because it hides the errorgroup. 

Can't we make use of the fact that the runner knows the context and when it is done anyways?
it could provide an interface to register parallelly executed teardown code.

```go
// rough pseudo-go
func httpService(runnner any)error{
httpserver :=server.new()
// will get executed automatically in a gouroutine when errgroup ctx is done
// (as soon as any runner routine errors)
runner.teardownfn(httpserver.shutdown)

// blocking call, but the outer function will be executed in a errorgroup.go.
// pass the errgroups ctx in case the function relies on ctx cancellation.
// explicitly getting the context from the runner makes it clear that this
// will be cancelled when other routines in the runner error.
// it could also implement the context interface itself
return httpserver.listen(runner.Context())
}
```
But I haven't thought this through too much

Reply:
We could implement that. But we would also still need the Defer functionality, (e.g. for closing database pools, which may be in use as long as one of the go routines is still alive).
We could also just pass the  errorgroup's context to the httpserver and let it shut down automatically instead of trying to do it gracefully.